plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/sampler/InferredActionSampler.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing.sampler;
+
+import org.opensearch.common.settings.Settings;
+import org.opensearch.telemetry.TelemetrySettings;
+import org.opensearch.telemetry.tracing.attributes.SamplingAttributes;
+import org.opensearch.telemetry.tracing.samplingResult.OTelSamplingResult;
+
+import java.util.List;
+import java.util.Objects;
+
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.trace.data.LinkData;
+import io.opentelemetry.sdk.trace.samplers.Sampler;
+import io.opentelemetry.sdk.trace.samplers.SamplingResult;
+
+/**
+ * InferredActionSampler implements a probability sampling strategy with sampling ratio as 1.0.
+ */
+public class InferredActionSampler implements Sampler {
+
+    private final Sampler fallbackSampler;
+    private final TelemetrySettings telemetrySettings;
+    private final Settings settings;
+
+    /**
+     * Constructor
+     *  @param telemetrySettings the telemetry settings
+     * @param settings          the settings
+     * @param fallbackSampler   the fallback sampler
+     */
+    private InferredActionSampler(TelemetrySettings telemetrySettings, Settings settings, Sampler fallbackSampler) {
+        this.telemetrySettings = Objects.requireNonNull(telemetrySettings);
+        this.settings = Objects.requireNonNull(settings);
+        this.fallbackSampler = fallbackSampler;
+    }
+
+    /**
+     * Create Inferred sampler.
+     *
+     * @param telemetrySettings the telemetry settings
+     * @param settings          the settings
+     * @param fallbackSampler   the fallback sampler
+     * @return the inferred sampler
+     */
+    public static Sampler create(TelemetrySettings telemetrySettings, Settings settings, Sampler fallbackSampler) {
+        return new InferredActionSampler(telemetrySettings, settings, fallbackSampler);
+    }
+
+    @Override
+    public SamplingResult shouldSample(
+        Context parentContext,
+        String traceId,
+        String name,
+        SpanKind spanKind,
+        Attributes attributes,
+        List<LinkData> parentLinks
+    ) {
+        boolean inferredSamplingAllowListed = telemetrySettings.getInferredSamplingAllowListed();
+        if (inferredSamplingAllowListed) {
+            Attributes customSampleAttributes = Attributes.builder()
+                .put(SamplingAttributes.SAMPLER.getValue(), SamplingAttributes.INFERRED_SAMPLER.getValue())
+                .putAll(attributes)
+                .build();
+            SamplingResult result = SamplingResult.recordAndSample();
+            return new OTelSamplingResult(result.getDecision(), customSampleAttributes);
```

Comment:
Sorry, I don't understand what this sampler is doing: it is 100% (aka always) sampler?

Reply:
Yes this is 100% sampler. We are using batch span processor itself for the final export so that will only export the span if the span is marked to be sampled. This takes care of marking all span as sampled so that the spans can be exported. With this it also adds an attribute to the span, for the processor layer to know that the span has been sampled by this span processor. 

Adding further, we have created a custom delegated processor layer which act as an intermediate layer between batch span processor and Otel framework and forwards the normal span i.e. spans which are not marked by this processor directly to batch span processor and if the span is marked by this sampler then filter out spans that are outlier or their child are outlier. 

Also adding further we need to mark the spans to be sampled beforehand so that the span can record information as well. If the span is not marked for sampling then it will not record any information as well and will act as a Noop span.